### DISCUSSION TITLE
Client Contract Boost Dashboard

### DISCUSSION THREAD

**Discussion by hammertoe** [filecoin-project/community#659] [Developers]
Hi folks, one of the problems that developers (especially during hackathons) were facing when trying to use the [Deal Client Contract in the Hardhat starter kit](https://github.com/filecoin-project/fevm-hardhat-kit/blob/main/contracts/basic-deal-client/DealClient.sol) was seeing where things were going wrong. 

This contract emits an event that can be listened to by Boost instances to instruct them to start a new storage deal with some data at a particular URL. The problem being that there are a number of parameters needed to be set, e.g. start epoch, duration, price, etc. and if any are incorrect then the Boost instance will reject the deal and the developer (and smart contract) have no visibility of why it was rejected.

PikNik, a storage provider that is running a miner on the Calibration testnet has created a dashboard allowing visibility into Boost logs to enable developers to diagnose why their deals might not be being accepted.

This dashboard is at: [http://38.70.220.87:8123/](http://38.70.220.87:8123/) and allows you to search by the client address (e.g. the address of the smart contract) and see the Boost log entires for it. 

Here is a step by step example of using it to debug deals using the client contract.

1. Install the [FEVM Hardhat starter kit](https://github.com/filecoin-project/fevm-hardhat-kit/tree/main). Once installed you can deploy the contracts as per the README in the starter kit. 

    ```
    matt@Matts-MBP fevm-hardhat-kit % yarn hardhat deploy 
    yarn run v1.22.19
    warning ../package.json: No license field
    $ /Users/matt/Development/fevm-hardhat-kit/node_modules/.bin/hardhat deploy
    Compiled 1 Solidity file successfully
    Wallet Ethereum Address: 0xCB091B88c7a166896FB946B236fed4CF4a747DD8
    deploying "SimpleCoin" (tx: 0x4418d766fd84fab005a69bce2f92cad02db087225dac0f836cd1e73f6dbeffd8)...: deployed at     0x6Da6F07751fBd8cbB2b50a74A32B267e0478ca8a with 44884936 gas
    deploying "FilecoinMarketConsumer" (tx: 0x62131c101d4d9f0291c2ef30af405be1985850357634d6c62b09ccba4ababa9d)...:     deployed at 0x6824224157Df4E8ae637758e2Dcb2f38d615d929 with 74496799 gas
    deploying "DealRewarder" (tx: 0x510b11e0e7e01d2e587f1b166c35d73819090ddd4b203a62e2ce782ac4127be0)...: deployed at 0x797Afb7DcD099104A8393AD1427F4b3cf3fef6D3 with 74154466 gas
    deploying "DealClient" (tx: 0xd3c99af9c2b2eac615f9125cf085725774cd8a8c58d6b93036cefe927df10577)...: deployed at 0xd91e5c247d4A450bAa83a4FBEE7691fF956491DF with 152596323 gas
    ```

    Note down the address the `DealClient` contract is deployed at. In my case it is `0xd91e5c247d4A450bAa83a4FBEE7691fF956491DF`.

2. Fund the contact from the [Calibration net tFIL faucet](https://faucet.calibration.fildev.network/funds.html) enter the contract address from above.

    <img width="1057" alt="image" src="https://github.com/filecoin-project/community/assets/1651303/9812f7dc-cd94-44fe-894d-368cd9684b87">

3. Make a deal using the `make-deal-proposal` task in the kit. Put in the contract address, and the other fields you can get from uploading a file to the [Lighthouse Data Depot](https://data.lighthouse.storage/), which will create a CAR file for you and give you all the details needed e.g.

    <img width="1009" alt="image" src="https://github.com/filecoin-project/community/assets/1651303/1a8715c8-194e-45c7-a9cd-4b52faf3315e">

    ```
    matt@Matts-MBP fevm-hardhat-kit % yarn hardhat make-deal-proposal \
      --network calibrationnet \                                                             
      --contract 0xd91e5c247d4A450bAa83a4FBEE7691fF956491DF \
      --piece-cid baga6ea4seaqeqtvkwvn42eincamnu7scmbeoifcqehurc3mgluzvlrosxx43qga \
      --piece-size 524288 \ 
      --verified-deal false \
      --label bafybeid7agv7gvah7vsdeaot6vy5j4jcy4hptufyjqdookzhefmmbdooxe \
      --start-epoch 869669 \
      --end-epoch 1471589 \
      --storage-price-per-epoch 4000000 \
      --provider-collateral 0 \
      --client-collateral 0 \
      --extra-params-version 1 \
      --location-ref "https://data-depot.lighthouse.storage/api/download/download_car?fileId=24945bf3-0859-4486-8c31-8eb03d4aa652.car" \
      --car-size 503736 \         
      --skip-ipni-announce false \
      --remove-unsealed-copy false
    yarn run v1.22.19
    warning ../package.json: No license field
    $ /Users/matt/Development/fevm-hardhat-kit/node_modules/.bin/hardhat make-deal-proposal --network calibrationnet --contract 0xd91e5c247d4A450bAa83a4FBEE7691fF956491DF --piece-cid baga6ea4seaqeqtvkwvn42eincamnu7scmbeoifcqehurc3mgluzvlrosxx43qga --piece-size 524288 --verified-deal false --label bafybeid7agv7gvah7vsdeaot6vy5j4jcy4hptufyjqdookzhefmmbdooxe --start-epoch 869669 --end-epoch 1471589 --storage-price-per-epoch 4000000 --provider-collateral 0 --client-collateral 0 --extra-params-version 1 --location-ref 'https://data-depot.lighthouse.storage/api/download/download_car?fileId=24945bf3-0859-4486-8c31-8eb03d4aa652.car' --car-size 503736 --skip-ipni-announce false --remove-unsealed-copy false
    Making deal proposal on network calibrationnet
    Complete! Event Emitted. ProposalId is: 0x6892b12a97899d784db3fd432368d77a7148bd5ab5e84dc2297298b7b5a3da3a
    ✨  Done in 92.24s.
    ```

5. Convert the contract address from `0x` format to `t410` format. The easiest way to do this is look the `0x` address up in the [Filfox calibration block explorer](https://calibration.filfox.info/en).

    <img width="1182" alt="image" src="https://github.com/filecoin-project/community/assets/1651303/841d53db-a77b-488e-bd37-2bc872246b63">

6. Enter the `t410` contract address into the [PikNik Boost looking glass](http://194.182.34.236:8123/). See the results. E.g.

    <img width="1211" alt="image" src="https://github.com/filecoin-project/community/assets/1651303/092495a1-f766-492d-869a-769903ecf56e">

In the case of a successful deal, you will see that shown in the logs, e.g.

```
2023-08-29T22:50:29.827+0200	INFO	boost-provider	storagemarket/contract_deal_monitor.go:180	received contract deal proposal	{"id": "0xb4ae50cdc6d78d898971e5c229fc77cfaf418da474c5d9a838b6a829285e1edb", "uuid": "161dfed9-3ad7-4044-a0be-522e5bda9898", "client-peer": "f410f3epfyjd5jjcqxkudut5645ur76kwjeo7c4aqlzq", "contract": "0xd91e5c247d4a450baa83a4fbee7691ff956491df", "piece-cid": "baga6ea4seaqldxiidebaup3ngh7t3pln6fadm3zsssxzo77fezuffxodhsgx2ja"}
2023-08-29T22:50:29.944+0200	INFO	boost-provider	storagemarket/contract_deal_monitor.go:189	contract deal proposal accepted	{"id": "0xb4ae50cdc6d78d898971e5c229fc77cfaf418da474c5d9a838b6a829285e1edb", "uuid": "161dfed9-3ad7-4044-a0be-522e5bda9898"}
```

If you get something wrong in the parameters you sent, you should see the error in the logs when the deal is rejected by Boost. e.g. in this case I got the duration slightly too short:

```
2023-08-29T19:32:30.182+0200	WARN	boost-provider	storagemarket/contract_deal_monitor.go:191	contract deal proposal rejected	{"id": "0x3e61bbd351be40198db73f3013bb201d78d9ec54ba0ea0a724dcaa78c962a16e", "uuid": "96989521-3298-47a6-adb0-d52d7a16efb1", "err": null, "reason": "failed validation: deal duration out of bounds (min, max, provided): 518400, 1555200, 515520"}
```

In the following case, I had not funded the smart contract, so it couldn't pay for the deal:

```
2023-08-29T21:21:00.413+0200	WARN	boost-provider	storagemarket/contract_deal_monitor.go:191	contract deal proposal rejected	{"id": "0x6ee5e47196d67bcb9c55bf9fe7064a79160da5bcb0b81b13281e338c2e79c214", "uuid": "67d06254-9493-4ca1-a6c4-a2ff6898894d", "err": null, "reason": "failed validation: client available funds in escrow 0 not enough to meet storage cost for deal 2407680000000"}
```

Hopefully this helps you if you are building using the Deal Client contract as a based.

**Comment by sumit03guha** [comment]
The given link `http://194.182.34.236:8123/` is not working.

**Reply by CubikRuubik** [reply]
Hi @sumit03guha, idk if still relevant but new one is http://38.70.220.87:8123/
